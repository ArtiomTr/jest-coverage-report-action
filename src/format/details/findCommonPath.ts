import { dirname } from 'path';

/**
 * Find a common path from a list of filepaths.
 */
export const findCommonPath = (filepaths: string[]): string => {
    let commonRoot = '';
    if (filepaths.length) {
        // If the paths are sorted, any prefix common to all paths will be common to the sorted first and last strings.
        const sortedPaths = [...filepaths].sort();
        const first = sortedPaths[0];
        const last = sortedPaths[sortedPaths.length - 1];

        const len = Math.min(first.length, last.length);
        for (let i = 0; i < len; i++) {
            const ch1 = first[i];
            const ch2 = last[i];
            if (ch1 == ch2) {
                commonRoot += ch1;
            } else {
                break;
            }
        }

        // If it doesn't appear to be a directory (partial filename), get the parent directory
        if (commonRoot.length && commonRoot[commonRoot.length - 1] !== '/') {
            commonRoot = dirname(commonRoot) + '/';
        }
    }
    return commonRoot;
};
