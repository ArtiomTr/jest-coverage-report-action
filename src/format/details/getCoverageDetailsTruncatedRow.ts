import { i18n } from '../../utils/i18n';

export const getCoverageDetailsTruncatedRow = (
    rowCount: number
): Array<string> => {
    return [
        '',
        i18n('detailsTruncatedRows', { truncatedRows: rowCount }),
        '',
        '',
        '',
        '',
    ];
};
