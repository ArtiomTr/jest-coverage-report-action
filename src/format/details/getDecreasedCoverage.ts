import { CoverageDetail, CoverageDetailsMap } from '../../typings/Coverage';

const coverageLessThan = (first: CoverageDetail, second: CoverageDetail) =>
    first.statements < second.statements ||
    first.branches < second.branches ||
    first.functions < second.functions;

export const getDecreasedCoverage = (
    headDetails: CoverageDetailsMap,
    baseDetails: CoverageDetailsMap | undefined
) =>
    Object.keys(headDetails)
        .filter(
            (filename) =>
                headDetails[filename] &&
                baseDetails?.[filename] &&
                coverageLessThan(headDetails[filename], baseDetails[filename])
        )
        .reduce<{
            headDetails: CoverageDetailsMap;
            baseDetails: CoverageDetailsMap;
        }>(
            (acc, filename) => {
                acc.headDetails[filename] = headDetails[filename];
                acc.baseDetails[filename] = baseDetails![filename];
                return acc;
            },
            { headDetails: {}, baseDetails: {} }
        );
