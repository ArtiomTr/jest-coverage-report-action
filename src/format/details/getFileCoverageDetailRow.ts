import { shrinkLongPath } from './shrinkLongPath';
import { CoverageDetail } from '../../typings/Coverage';
import { formatPercentage } from '../../utils/formatPercentage';
import { getStatusOfPercents } from '../../utils/getStatusOfPercents';

export const getFileCoverageDetailRow = (
    filename: string,
    headDetail: CoverageDetail,
    baseDetail?: CoverageDetail,
    threshold?: number
): Array<string> => [
    getStatusOfPercents(headDetail.lines, threshold),
    shrinkLongPath(filename),
    formatPercentage(headDetail.statements, baseDetail?.statements),
    formatPercentage(headDetail.branches, baseDetail?.branches),
    formatPercentage(headDetail.functions, baseDetail?.functions),
    formatPercentage(headDetail.lines, baseDetail?.lines),
];
