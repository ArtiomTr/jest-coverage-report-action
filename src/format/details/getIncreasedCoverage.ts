import { CoverageDetail, CoverageDetailsMap } from '../../typings/Coverage';

const coverageGreaterThan = (first: CoverageDetail, second: CoverageDetail) =>
    first.statements > second.statements ||
    first.branches > second.branches ||
    first.functions > second.functions;

export const getIncreasedCoverage = (
    headDetails: CoverageDetailsMap,
    baseDetails: CoverageDetailsMap | undefined
) =>
    Object.keys(headDetails)
        .filter(
            (filename) =>
                headDetails[filename] &&
                baseDetails?.[filename] &&
                coverageGreaterThan(headDetails[filename], baseDetails[filename])
        )
        .reduce<{
            headDetails: CoverageDetailsMap;
            baseDetails: CoverageDetailsMap;
        }>(
            (acc, filename) => {
                acc.headDetails[filename] = headDetails[filename];
                acc.baseDetails[filename] = baseDetails![filename];
                return acc;
            },
            { headDetails: {}, baseDetails: {} }
        );
