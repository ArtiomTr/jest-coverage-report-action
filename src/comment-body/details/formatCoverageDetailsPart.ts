import markdownTable from 'markdown-table';

import { getFileCoverageDetailRow } from './getFileCoverageDetailRow';
import { ParsedCoverageDetails } from '../../collect-coverage/parseCoverageDetails';
import { details } from '../strings.json';
import { createMarkdownSpoiler } from '../utils/createMarkdownSpoiler';
import { formatHeadingAndTable } from '../utils/formatHeadingAndTable';

export type DetailsFormatOptions = {
    summary: string;
    heading: string;
};

export const formatCoverageDetailsPart = (
    formatOptions: DetailsFormatOptions,
    headDetails: ParsedCoverageDetails,
    baseDetails?: ParsedCoverageDetails
): string | undefined => {
    const { summary, heading } = formatOptions;

    const tableContent = Object.keys(headDetails).map((filename) =>
        getFileCoverageDetailRow(
            filename,
            headDetails[filename],
            baseDetails?.[filename]
        )
    );

    if (tableContent.length > 0) {
        return createMarkdownSpoiler({
            body: formatHeadingAndTable(
                heading,
                markdownTable([details.columnHeaders, ...tableContent])
            ),
            summary,
        });
    }

    return undefined;
};
