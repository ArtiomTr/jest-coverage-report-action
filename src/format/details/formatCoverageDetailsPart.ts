import markdownTable from 'markdown-table';

import { getFileCoverageDetailRow } from './getFileCoverageDetailRow';
import { CoverageDetailsMap } from '../../typings/Coverage';
import { createMarkdownSpoiler } from '../../utils/createMarkdownSpoiler';
import { i18n } from '../../utils/i18n';
import { withExplanation } from '../../utils/withExplanation';

export const formatCoverageDetailsPart = (
    summary: string,
    headDetails: CoverageDetailsMap,
    baseDetails?: CoverageDetailsMap,
    threshold?: number
): string | undefined => {
    const tableContent = Object.keys(headDetails).map((filename) =>
        getFileCoverageDetailRow(
            filename,
            headDetails[filename],
            baseDetails?.[filename],
            threshold
        )
    );

    if (tableContent.length > 0) {
        return createMarkdownSpoiler({
            body: markdownTable(
                [
                    [
                        withExplanation(
                            i18n('status'),
                            i18n('statusExplanation')
                        ),
                        i18n('filename'),
                        i18n('statements'),
                        i18n('branches'),
                        i18n('functions'),
                        i18n('lines'),
                    ],
                    ...tableContent,
                ],
                {
                    align: ['c', 'l', 'l', 'l', 'l', 'l'],
                }
            ),
            summary,
        });
    }

    return undefined;
};
