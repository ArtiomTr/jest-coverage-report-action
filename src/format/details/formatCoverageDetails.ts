import { formatCoverageDetailsPart } from './formatCoverageDetailsPart';
import { getDecreasedCoverage } from './getDecreasedCoverage';
import { getNewFilesCoverage } from './getNewFilesCoverage';
import { CoverageDetailsMap } from '../../typings/Coverage';
import { i18n } from '../../utils/i18n';

export const formatCoverageDetails = (
    headDetails: CoverageDetailsMap,
    baseDetails: CoverageDetailsMap | undefined,
    threshold: number | undefined
): string => {
    const decreasedCoverage = getDecreasedCoverage(headDetails, baseDetails);

    return [
        formatCoverageDetailsPart(
            i18n('newFilesCoverage'),
            getNewFilesCoverage(headDetails, baseDetails),
            undefined,
            threshold
        ),
        formatCoverageDetailsPart(
            i18n('decreasedCoverageFiles'),
            decreasedCoverage.headDetails,
            decreasedCoverage.baseDetails,
            threshold
        ),
    ].join('\n');
};
