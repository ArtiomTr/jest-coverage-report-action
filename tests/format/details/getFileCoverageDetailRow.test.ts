import { getFileCoverageDetailRow } from '../../../src/format/details/getFileCoverageDetailRow';

describe('getFileCoverageDetailRow', () => {
    it('should return formatted detail row (decreased coverage)', () => {
        expect(
            getFileCoverageDetailRow(
                'hello.ts',
                {
                    filename: 'hello.ts',
                    lines: 50,
                    branches: 100,
                    functions: 50,
                    statements: 50,
                },
                {
                    filename: 'hello.ts',
                    lines: 60,
                    branches: 100,
                    functions: 60,
                    statements: 60,
                },
                70
            )
        ).toMatchSnapshot();
    });

    it('should return formatted detail row (increased coverage)', () => {
        expect(
            getFileCoverageDetailRow(
                'hello.ts',
                {
                    filename: 'hello.ts',
                    lines: 50,
                    branches: 100,
                    functions: 50,
                    statements: 50,
                },
                {
                    filename: 'hello.ts',
                    lines: 40,
                    branches: 100,
                    functions: 40,
                    statements: 40,
                },
                70
            )
        ).toMatchSnapshot();
    });

    it('should return formatted detail row (no base details)', () => {
        expect(
            getFileCoverageDetailRow(
                'hello.ts',
                {
                    filename: 'hello.ts',
                    lines: 50,
                    branches: 100,
                    functions: 50,
                    statements: 50,
                },
                undefined,
                70
            )
        ).toMatchSnapshot();
    });
});
