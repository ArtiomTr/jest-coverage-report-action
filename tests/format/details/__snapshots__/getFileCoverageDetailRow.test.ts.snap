// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getFileCoverageDetailRow > should return formatted detail row (decreased coverage) 1`] = `
[
  "ğŸ”´",
  "hello.ts",
  "<div title=\\"Base coverage is: 60%\\">50% (-10% ğŸ”»)</div>",
  "100%",
  "<div title=\\"Base coverage is: 60%\\">50% (-10% ğŸ”»)</div>",
  "<div title=\\"Base coverage is: 60%\\">50% (-10% ğŸ”»)</div>",
]
`;

exports[`getFileCoverageDetailRow > should return formatted detail row (increased coverage) 1`] = `
[
  "ğŸ”´",
  "hello.ts",
  "<div title=\\"Base coverage is: 40%\\">50% (+10% ğŸ”¼)</div>",
  "100%",
  "<div title=\\"Base coverage is: 40%\\">50% (+10% ğŸ”¼)</div>",
  "<div title=\\"Base coverage is: 40%\\">50% (+10% ğŸ”¼)</div>",
]
`;

exports[`getFileCoverageDetailRow > should return formatted detail row (no base details) 1`] = `
[
  "ğŸ”´",
  "hello.ts",
  "50%",
  "100%",
  "50%",
  "50%",
]
`;
